internal/assertions/highlighting_assertion.go

```diff
@@ -0,0 +1,172 @@
+	a.matchesShouldbeHighlighted = strings.Contains(a.ExpectedOutput, "\033[01;31m")
+	a.logger = logger.Clone()
+	a.actualResult = result
+
+	expectedLines := utils.ProgramOutputToLines(a.ExpectedOutput)
+	actualLines := utils.ProgramOutputToLines(string(result.Stdout))
+
+	// The dimensions for the VT will be the maximum of the expected and actual output's width and height.
+	// Add 1 to the maximum length because, even in the case of empty input, we need to spawn a VT.
+	maxTerminalWidth := 1 + max(
+		len(a.ExpectedOutput),
+		len(result.Stdout),
+	)
+
+	maxTerminalHeight := 1 + max(
+		len(expectedLines),
+		len(actualLines),
+	)
+
+	virtualTerminal1 := virtual_terminal.NewCustomVT(maxTerminalHeight, maxTerminalWidth)
+	defer virtualTerminal1.Close()
+
+	virtualTerminal2 := virtual_terminal.NewCustomVT(maxTerminalHeight, maxTerminalWidth)
+	defer virtualTerminal2.Close()
+
+	if _, err := virtualTerminal1.WriteWithCrlfTranslation([]byte(a.ExpectedOutput)); err != nil {
+		return err
+	}
+
+	if _, err := virtualTerminal2.WriteWithCrlfTranslation(result.Stdout); err != nil {
+		return err
+	}
+
+	expectedScreenState := virtualTerminal1.GetScreenState()
+	a.panicIfExpectedScreenStateisFlawed(expectedScreenState)
+	actualScreenState := virtualTerminal2.GetScreenState()
+
+	// Assert text contents first
+	if err := a.assertTextContents(expectedScreenState, actualScreenState); err != nil {
+		return err
+	}
+
+	// Assert highlighting
+	return a.assertHighlighting(expectedScreenState, actualScreenState)
+}
+
+func (a *HighlightingAssertion) assertTextContents(expectedScreenState, actualScreenState *virtual_terminal.ScreenState) error {
+	expectedLines := expectedScreenState.GetLinesOfTextUptoCursor()
+	actualLines := actualScreenState.GetLinesOfTextUptoCursor()
+
+	orderedLinesAssertion := OrderedLinesAssertion{
+		ExpectedOutputLines: expectedLines,
+	}
+
+	// Simulate the result as if it would have been produced without highlighting it
+	actualOutput := strings.Join(actualLines, "\n")
+
+	actualResult := executable.ExecutableResult{
+		Stdout: []byte(actualOutput),
+	}
+
+	return orderedLinesAssertion.Run(actualResult, a.logger)
+}
+
+func (a *HighlightingAssertion) assertHighlighting(expectedScreenState, actualScreenState *virtual_terminal.ScreenState) error {
+	// Initialize comparison context
+	outputLinesCount := len(expectedScreenState.GetLinesOfTextUptoCursor())
+
+	// Hand off to comparator for comparison
+	screenStateComparator := newScreenStateComparator()
+	comparisonError := screenStateComparator.CompareHighlighting(expectedScreenState, actualScreenState)
+
+	lastSuccessFulRowIndex := outputLinesCount - 1
+	if comparisonError != nil {
+		lastSuccessFulRowIndex = comparisonError.RowIdx - 1
+	}
+
+	// Log success messages for each row
+	actualLines := actualScreenState.GetLinesOfTextUptoCursor()
+	for rowIdx := 0; rowIdx <= lastSuccessFulRowIndex; rowIdx++ {
+		if a.matchesShouldbeHighlighted {
```

Comment:
This value applies to the whole assertion, what if we had 3 lines where line 1 & 3 required highlighting but line 2 didn't? We'd incorrectly say "All matches in line 2 are highlighted". 

Can we use a local variable here instead like `hasExpectedMatches` or something like that?